### PR TITLE
Cache partner referrals signup links

### PR DESCRIPTION
**Pull request by Dinamiko**
Cache partner referrals signup links to avoid unnecessari API calls.

### CODE REVIEW

File: modules/ppcp-onboarding/src/Render/OnboardingRenderer.php [hunk 1]
```diff
@@ -83,9 +94,17 @@ public function get_signup_link( bool $is_production, array $products ) {
...
+		$environment = $is_production ? 'production' : 'sandbox';
+		$product     = isset( $data['products']['PPCP'] ) ? 'ppcp' : 'express-checkout';
+		if ( $this->cache->has( $environment . '-' . $product ) ) {
+			return $this->cache->get( $environment . '-' . $product );
+		}
+
 		$url = $is_production ? $this->production_partner_referrals->signup_link( $data ) : $this->sandbox_partner_referrals->signup_link( $data );
 		$url = add_query_arg( $args, $url );
 
+		$this->cache->set( $environment . '-' . $product, $url );
```
Comment (AlexP11223) [inline modules/ppcp-onboarding/src/Render/OnboardingRenderer.php]: It seems like `Cache` sets `0` for transient expiration. May cause problems if the link stops working after some time?

https://github.com/woocommerce/woocommerce-paypal-payments/blob/d4c7dc6b6ba6ae33741c9bdb9e2682352958fea1/modules/ppcp-api-client/src/Helper/Cache.php#L74

**Comment by Dinamiko** [comment]
@AlexP11223 How do we get rid of these errors? https://github.com/woocommerce/woocommerce-paypal-payments/runs/6607234216?check_suite_focus=true

I see `psalm-baseline.xml` includes `DAY_IN_SECONDS` for example, but the file seems autogenerated, how does it work? thanks.

**Comment by AlexP11223** [comment]
baseline is for ignoring old errors that were there before adding psalm.

You can add to https://github.com/woocommerce/woocommerce-paypal-payments/blob/trunk/.psalm/stubs.php

File: modules/ppcp-onboarding/src/Render/OnboardingRenderer.php [hunk 2]
```diff
@@ -83,9 +94,17 @@ public function get_signup_link( bool $is_production, array $products ) {
...
+		$environment = $is_production ? 'production' : 'sandbox';
+		$product     = isset( $data['products']['PPCP'] ) ? 'ppcp' : 'express-checkout';
+		if ( $this->cache->has( $environment . '-' . $product ) ) {
+			return $this->cache->get( $environment . '-' . $product );
+		}
+
 		$url = $is_production ? $this->production_partner_referrals->signup_link( $data ) : $this->sandbox_partner_referrals->signup_link( $data );
 		$url = add_query_arg( $args, $url );
 
+		$this->cache->set( $environment . '-' . $product, $url, 3 * MONTH_IN_SECONDS );
```
Comment (AlexP11223) [inline modules/ppcp-onboarding/src/Render/OnboardingRenderer.php:106]: 3 months seems a lot, is that how long the link supposed to work?

Comment (Dinamiko) [reply modules/ppcp-onboarding/src/Render/OnboardingRenderer.php:106]: yup, I asked PayPal and they said that referral link should be valid for about 3 months.

**Review by AlexP11223 — APPROVED** [review]

**Pull request merged**